### PR TITLE
fix: Track node labels propagated to the endpoint manager correctly

### LINKED ISSUE #29649: Node labels updates rejected indefinitely by host endpoint if node watcher initializes before the host endpoint is created
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We ran into a race condition where the host endpoint's labels are outdated indefinitely. The issue is fixed if I restart the cilium agent.

The sequence of steps that resulted in the race condition are:

1. The endpoint manager's node label observer is initialized before the host endpoint is created. So, the first node label update is missed as we return early if the host endpoint is not found.

https://github.com/cilium/cilium/blob/d1128cf948979de7433385b2101a98404d99c64f/pkg/endpointmanager/host.go#L43-L48

2. All the subsequent node label updates are rejected by endpoint's `ModifyIdentityLabels` function

https://github.com/cilium/cilium/blob/d1128cf948979de7433385b2101a98404d99c64f/pkg/labels/oplabels.go#L157-L169

As the first node label update was missed, some labels on the oldNode object are not found on the endpoint's existing labels so the update request is rejected.

### Cilium Version

All versions

### Kernel Version

Linux bootstrapper 5.4.0-100-generic

### Kubernetes Version

1.27

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by skmatti**
This fixes a race condition during agent startup that causes the k8s node label updates to not get propagated to the host endpoint.

The current endpoint label update logic rejects a request if any of the labels on the old node are not present in the endpoint manager state.

For host endpoints, the k8s node label add/update events may be missed if the k8s node watcher initializes before the host endpoint is created. As a result, the host endpoint labels are outdated, and all subsequent node label updates are rejected due to the aforementioned precondition check.

To resolve the issue, this commit updates the old labels only if the current node label update is successfully propagated to the endpoint manager.

Fixes cilium#29649

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

Fixes: #issue-number

```release-note
Fixes a race condition during agent startup that causes the k8s node label updates to not get propagated to the host endpoint.
```

### CODE REVIEW

**Comment by skmatti** [comment]
Most of the failures are context deadline exceed installing cilium

**Comment by dylandreimerink** [comment]
We managed to cause rate limits in both the github API and docker registry. Retries should solve these, triggering these now

**Comment by skmatti** [comment]
Thanks, all the tests are passing now. @pchaigno can you take a look?

**Comment by pchaigno** [comment]
> Thanks, all the tests are passing now.

Yep, we're just waiting for the review from @squeed.

**Review by squeed — APPROVED** [review]
neat and clean. Thanks!

**Comment by skmatti** [comment]
Thanks @squeed. Can we mark this change for backporting to 1.16 and 1.15?.

**Comment by joestringer** [comment]
@skmatti It looks like the patch did not apply cleanly. Would you be open to preparing a port of the change onto the v1.15 branch?

I've also tagged it for v1.14 as I think that makes sense given that this can impact the labels associated with the host.

**Comment by joestringer** [comment]
I realized I'm actually not sure whether this affects 1.13/.14. Let me know if you figure that out and I can set the labels appropriately.

**Comment by skmatti** [comment]
Yeah, this definitely affects all the versions. I can open a change against 1.15. The backport to 1.14 and 1.13 will be slightly different as there are node watch logic changed between 1.14 and 1.15.

**Comment by giorio94** [comment]
Backported to v1.14 via https://github.com/cilium/cilium/pull/31735.

**Pull request merged**
